neural_nets/losses/losses.py

```diff
@@ -385,3 +385,66 @@ def grad(self, Y_fake, module, Y_real=None, gradInterp=None):
         else:
             raise ValueError("Unrecognized module: {}".format(module))
         return grad
+
+
+class Cosine(ObjectiveBase):
+    def __init__(self):
+        super().__init__()
+
+    def __call__(self, y, y_pred):
+        return self.loss(y, y_pred)
+
+    def __str__(self):
+        return "Cosine Proximity"
+
+    @staticmethod
+    def loss(y, y_pred):
+        """
+        Compute the Cosine distance between 1-D arrays.
+
+        The Cosine distance between `u` and `v`, is defined as
+        .. math::
+            1 - \\frac{u \\cdot v}
+                      {||u||_2 ||v||_2}.
+        where :math:`u \\cdot v` is the dot product of :math:`u` and
+        :math:`v`.
+
+        Parameters
+        ----------
+        y : numpy array of shape (n, m)
+
+        y_pred : numpy array of shape (n, m)
+
+
+        Returns
+        -------
+        loss : float
+
+        """
+        def l2_normalize(x, axis=-1):
+            y = np.max(np.sum(x ** 2, axis, keepdims=True), axis, keepdims=True)
+            return x / np.sqrt(y)
+
+        y_true = l2_normalize(y, axis=-1)
+        y_pred = l2_normalize(y_pred, axis=-1)
+        return 1. - np.sum(y_true * y_pred, axis=-1)
```

Comment:
I saw 2 different implementation. First one is 

```python
return -np.sum(y_true * y_pred, axis=-1)
```

or 

```python
return 1. - np.sum(y_true * y_pred, axis=-1)
```

Which one should we choose for our implementation? They are same in nature.

Reply:
I like the first since it ranges between -1 and 1, similar to the cosine distance itself.